### ISSUE TITLE
Should we name all the pattern-matching lambdas?

### ISSUE THREAD

**Issue by gallais** [agda/agda-stdlib#1166]
Remember that two similar-looking pattern-matching lambdas are not judgmentally
equal because each is elaborated to a different definition with a different
internal name.

One possible workaround is to introduce a named where-clause that allows us to
pick a name for this anonymous lambda. If we adopt a clear naming scheme for
these, we could have very predictable names for users.

Using `Codata.Thunk`'s `cofix` as an example, we could write:

```diff
  cofix : ∀[ Thunk P ⇒ P ] → ∀[ P ]
-  cofix f = f (λ where .force → cofix f)
+  cofix f = f aux module Cofix where
+    aux = λ where .force → cofix f
```

And we would then be able to mention the anonymous lambda by its name
`Cofix.aux`.

With this scheme, we could write some of the expressions that show up in the
bisimilarity reasoning in e.g. [`Codata.Stream.Properties`](http://agda.github.io/agda-stdlib/Codata.Stream.Properties.html#2051) that have to be left
blank at the moment.

Pros: we finally have names for these
Cons: we pollute the name space with a lot of module names

**Comment by MatthewDaggitt** [comment]
I vaguely remember it being said in the Agda issue discussing this that the fact that identical lambdas had different names was considered a bug? 

Unless there's places where the lambda's can't be inferred and that's causing problems, I'd probably err on the side of leaving as is and waiting for a fix upstream. As you say, it adds a lot of pollution to the namespace, for a comparatively small(?) increase in readability...

**Comment by jespercockx** [comment]
I don't think it is considered a bug. Having structural equality of pattern lambdas would certainly be possible in theory but it would require a lot of work to implement and it would come with a certain performance penalty, as comparing two definitions is definitely more expensive than merely comparing two names. So I don't expect there will be an upstream fix anytime soon.

**Comment by JacquesCarette** [comment]
My reaction, even after a few days of thinking about it, is still negative.

If a sub-function is really needed elsewhere, it should be given a real name.  [For `cofix`, you'd use a `mutual` block].

**Comment by gallais** [comment]
The problem with the mutual blocks is that you need to specify the type of
the auxiliary function. At this point you have to declare all coinductive
function twice.

**Comment by JacquesCarette** [comment]
True... but perhaps that's still a good thing? In the sense that if you wish to expose certain 'innards', then there is a cost, of an extra type signature. Perhaps that should cause you to pause and wonder if you truly want to expose that much.

**Comment by MatthewDaggitt** [comment]
@gallais I'm more than happy to defer to your and @JacquesCarette's decision on this as I've never really used the `Codata` part of the library...

**Comment by jamesmckinna** [comment]
After my (still limited) experience with `Data.List.Base.{inits|tails}` and `Data.List.Scans.*` and friends, I'm generally *pro* the suggestion, because the 'cost' of namespace pollution seems to be confined to having a module name be (a version of) the capitalised name of the function name being defined, which is both cognate/intelligible/legible/discoverable, and no worse than the name of the function being defined in the first place.

Doubtless there'll be exceptions/corner cases, but this on balance seems a good idea, to me at least!

**Comment by JacquesCarette** [comment]
This is indeed a naming issue, not a `Codata` issue. I have now seen lots of uses of the local `module` pattern that @gallais uses above in `1lab`, and I've warmed up to it a lot. While naming will still continue to be a pain, I would be in favour of naming more things.

**Comment by jamesmckinna** [comment]
Regarding the actual name(s) chosen:
* the library as it stands currently, seems to adopt the haskell `go` convention, but in places the name `helper` is used
* @gallais above offers `aux`, which once upon a time, I would have chosen/preferred (there's maybe some Latin-derived vs. German-derived influence in the background here)
* in the examples I cite, because the auxiliary (sic!) function is computing the list of tails of the list(s) at hand, I chose
  - `tails` for the `aux` name (so: a 'semantic'/mnemonic name, maybe that's a bad idea?)
  - module name follows top-level function name, capitalised, as with `Cofix` above also

What names *should* we agree on as `style-guide` policy?
